OpenCV 3.1 build with QT and Cuda 8.0 errors

<!--
If you have a question rather than reporting a bug please go to ma where you get much faster responses.
If you need further assistance please read [How To Contribute](https://github.com/opencv/opencv/wiki/How_to_contribute).

This is a template helping you to create an issue which can be processed as quickly as possible. This is the bug reporting section for the OpenCV library.
-->

##### System information (version)
<!-- Example
- OpenCV => 3.1
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2015
-->

- OpenCV => 3.1
- Operating System / Platform => Ubuntu 16.04
- Compiler => gcc/g++ 5.4

##### Detailed description

<!-- your description -->

I try to build opencv 3.1 from master branch, but I have errors: 

/home/ubuntu/opencv/modules/highgui/src/window_QT.cpp: In member function ‘virtual void OpenGlViewPort::wheelEvent(QWheelEvent*)’:
/home/ubuntu/opencv/modules/highgui/src/window_QT.cpp:3204:32: error: ‘scaleView’ was not declared in this scope
     scaleView(delta / 240.0, pt);

/home/ubuntu/opencv/modules/highgui/src/window_QT.cpp:3205:14: error: ‘viewport’ was not declared in this scope
     viewport()->update();

Also I had error:

In file included from /home/ubuntu/opencv/build/modules/cudalegacy/opencv_test_cudalegacy_pch_dephelp.cxx:1:0:
/home/ubuntu/opencv/modules/cudalegacy/test/test_precomp.hpp:54:69: error: operator '&&' has no right operand

I fixed it by removing "\n" in test_precomp.hpp on line 54.

If I try to build without QT, I get errors:

/usr/local/cuda-8.0/include/thrust/detail/reference.inl(127): error: no default constructor exists for class "thrust::detail::execute_with_allocator<cv::cuda::device::ThrustAllocator, thrust::system::cuda::detail::execute_on_stream_base>"

/usr/local/cuda-8.0/include/thrust/detail/reference.inl(98): error: no default constructor exists for class "thrust::detail::execute_with_allocator<cv::cuda::device::ThrustAllocator, thrust::system::cuda::detail::execute_on_stream_base>"

##### Steps to reproduce

Clone from https://github.com/opencv/opencv

Run cmake with -D CMAKE_BUILD_TYPE=RELEASE -D CMAKE_INSTALL_PREFIX=/usr/local -D WITH_TBB=ON -D WITH_V4L=ON -D WITH_QT=ON -D WITH_OPENGL=ON or without WITH_QT=ON

Run make

--- a/modules/cudalegacy/test/TestHaarCascadeApplication.cpp
+++ b/modules/cudalegacy/test/TestHaarCascadeApplication.cpp
@@ -52,8 +52,7 @@ namespace
         ~FpuControl();
 
     private:
-    #if defined(__GNUC__) && !defined(__APPLE__) && !defined(__arm__) &&
-        !defined(__aarch64__) && !defined(__powerpc64__)
+    #if defined(__GNUC__) && !defined(__APPLE__) && !defined(__arm__) && !defined(__aarch64__) && !defined(__powerpc64__)
         fpu_control_t fpu_oldcw, fpu_cw;
     #elif defined(_WIN32) && !defined(_WIN64)
         unsigned int fpu_oldcw, fpu_cw;
@@ -62,8 +61,7 @@ namespace
 
     FpuControl::FpuControl()
     {
-    #if defined(__GNUC__) && !defined(__APPLE__) && !defined(__arm__) &&
-        !defined(__aarch64__) && !defined(__powerpc64__)
+    #if defined(__GNUC__) && !defined(__APPLE__) && !defined(__arm__) && !defined(__aarch64__) && !defined(__powerpc64__)
         _FPU_GETCW(fpu_oldcw);
         fpu_cw = (fpu_oldcw & ~_FPU_EXTENDED & ~_FPU_DOUBLE & ~_FPU_SINGLE) | _FPU_SINGLE;
         _FPU_SETCW(fpu_cw);
@@ -76,8 +74,7 @@ namespace
 
     FpuControl::~FpuControl()
     {
-    #if defined(__GNUC__) && !defined(__APPLE__) && !defined(__arm__) &&
-        !defined(__aarch64__) && !defined(__powerpc64__)
+    #if defined(__GNUC__) && !defined(__APPLE__) && !defined(__arm__) && !defined(__aarch64__) && !defined(__powerpc64__)
         _FPU_SETCW(fpu_oldcw);
     #elif defined(_WIN32) && !defined(_WIN64)
         _controlfp_s(&fpu_cw, fpu_oldcw, _MCW_PC);

--- a/modules/cudalegacy/test/test_precomp.hpp
+++ b/modules/cudalegacy/test/test_precomp.hpp
@@ -51,8 +51,7 @@
 #ifndef __OPENCV_TEST_PRECOMP_HPP__
 #define __OPENCV_TEST_PRECOMP_HPP__
 
-#if defined(__GNUC__) && !defined(__APPLE__) && !defined(__arm__) &&
-    !defined(__aarch64__) && !defined(__powerpc64__)
+#if defined(__GNUC__) && !defined(__APPLE__) && !defined(__arm__) && !defined(__aarch64__) && !defined(__powerpc64__)
     #include <fpu_control.h>
 #endif
 

